core/src/main/java/org/apache/calcite/util/graph/AttributedDirectedGraph.java

```diff
@@ -57,20 +57,21 @@ public E addEdge(V vertex, V targetVertex) {
   }
 
   public E addEdge(V vertex, V targetVertex, Object... attributes) {
-    final VertexInfo<V, E> info = vertexMap.get(vertex);
-    if (info == null) {
+    final VertexInfo<V, E> outInfo = vertexOutMap.get(vertex);
```

Comment:
Looks like addEdge(), removeEdge() has the same logics as the base class. Why do we need to override here?

Reply:
The only difference is the method for creating edges. In the subclass, some attributes are used to create the edge, and the attributes are passed in from the method parameter list. This makes the method signatures different, so we cannot use the methods from the super class.

Maybe we can solve this problem in a future issue.